### PR TITLE
refactor: pipeline reconciler

### PR DESCRIPTION
**Pull request by sfwn**
#### What this PR does / why we need it:

refactor pipeline reconciler

#### Which issue(s) this PR fixes:

- [Erda Cloud Issue Link](https://erda.cloud/erda/dop/projects/387/issues/all?id=290594&issueFilter__urlQuery=eyJzdGF0ZXMiOls0NDA2LDQ0MDddfQ%3D%3D&issueTable__urlQuery=eyJwYWdlTm8iOjEsICJwYWdlU2l6ZSI6MTB9&iterationID=1146&type=TASK)

#### Specified Reviewers:

/assign @Effet 

#### ChangeLog

| Language | Changelog |
| --------- | ------------ |
| 🇺🇸 English |  refactor pipeline reconciler            |
| 🇨🇳 中文    |  重构 pipeline reconciler            |

### CODE REVIEW

File: modules/pipeline/providers/reconciler/task_reconciler.go [hunk 1]
```diff
@@ -0,0 +1,391 @@
...
+	"github.com/erda-project/erda/pkg/loop"
+	"github.com/erda-project/erda/pkg/strutil"
+)
+
+type TaskReconciler interface {
+	// ReconcileOneTaskUntilDone reconcile one task until done.
+	// done means end status, include success, failed and others.
+	ReconcileOneTaskUntilDone(ctx context.Context, p *spec.Pipeline, task *spec.PipelineTask)
+
+	SaveTaskIdempotently(ctx context.Context, p *spec.Pipeline, task *spec.PipelineTask) error
```
Comment (Effet) [inline modules/pipeline/providers/reconciler/task_reconciler.go:45]: is `IdempotentSaveTask` more proper? @shuofan

**Review by Effet — CHANGES_REQUESTED** [review]

File: modules/pipeline/providers/reconciler/cache/impl.go [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+	"github.com/erda-project/erda/modules/pipeline/spec"
+	"github.com/erda-project/erda/pkg/parser/pipelineyml"
+)
+
+const (
+	pipelineStagesContextCachesPrefixKey               = "reconciler_caches_stages"
+	pipelineYmlContextCachesPrefixKey                  = "reconciler_caches_yml"
+	pipelineRerunSuccessTaskContextCachesPrefixKey     = "reconciler_caches_rerun_success_tasks"
+	pipelinePassedDataWhenCreateContextCachesPrefixKey = "reconciler_caches_passed_data_when_create"
+
```
Comment (Effet) [inline modules/pipeline/providers/reconciler/cache/impl.go:32]: why leave a blank line here?

File: modules/pipeline/providers/reconciler/pipeline_reconciler.go [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+	// UpdateCurrentReconcileStatusIfNecessary calculate current reconcile status and update if necessary.
+	UpdateCurrentReconcileStatusIfNecessary(ctx context.Context, p *spec.Pipeline) error
+
+	// TeardownAfterReconcileDone teardown one pipeline after reconcile done.
+	TeardownAfterReconcileDone(ctx context.Context, p *spec.Pipeline)
+}
+
+type defaultPipelineReconciler struct {
+	r *provider
+
```
Comment (Effet) [inline modules/pipeline/providers/reconciler/pipeline_reconciler.go:63]: ugly blank lines

File: modules/pipeline/providers/reconciler/reconcile.go [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+func (r *provider) ReconcileOnePipeline(ctx context.Context, pipelineID uint64) {
+	// make pipelineID-level context to support cancel reconcile
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	defer func() {
+		if err := recover(); err != nil {
+			r.Log.Errorf("panic while reconcile one pipeline until done, cancel reconcile, pipelineID: %d, err: %v", pipelineID, err)
+			debug.PrintStack()
+		}
+	}()
```
Comment (Effet) [inline modules/pipeline/providers/reconciler/reconcile.go:37]: I think cancel do first is better? if panic

File: modules/pipeline/providers/reconciler/rutil/continue_working.go [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+	"fmt"
+	"time"
+
+	"github.com/erda-project/erda-infra/base/logs"
+)
+
+type WaitDuration time.Duration
+
+var (
+	NoNeedContinueWorking              WaitDuration = -3
```
Comment (Effet) [inline modules/pipeline/providers/reconciler/rutil/continue_working.go:28]: better to unify them? `ContinueWorkingNoNeed`

File: modules/pipeline/providers/reconciler/task_reconciler.go [hunk 6]
```diff
@@ -0,0 +1,391 @@
...
+	rutil.ContinueWorking(ctx, tr.log, func(ctx context.Context) rutil.WaitDuration {
+		// get executor
+		executor, err := actionexecutor.GetManager().Get(types.Name(task.GetExecutorName()))
+		if err != nil {
+			tr.log.Errorf("failed to get task executor(auto retry), err: %v", err)
+			return rutil.ContinueWorkingWithCustomInterval(tr.pr.r.Cfg.RetryInterval)
+		}
+
+		// generate framework to run task
+		framework := taskrun.New(ctx, task, executor, p, tr.pr.r.bdl, tr.pr.r.dbClient, tr.pr.r.actionAgentSvc, tr.pr.r.extMarketSvc, tr.pr.r.ClusterInfo)
```
Comment (Effet) [inline modules/pipeline/providers/reconciler/task_reconciler.go:182]: ugly params, can you refactor to a better New?

File: modules/pipeline/providers/cron/compensator/provider.go [hunk 7]
```diff
@@ -186,7 +230,7 @@ func (p *provider) cronInterruptCompensate(pc spec.PipelineCron) error {
 
 	if len(needTriggerTimes) > 0 {
 		// Pipeline search record created based on createlyname + source
-		existPipelines, _, _, _, err := p.client.PageListPipelines(apistructs.PipelinePageListRequest{
```
Comment (Effet) [inline modules/pipeline/providers/cron/compensator/provider.go:189]: better to define a struct, too many returns

File: modules/pipeline/providers/reconciler/reconcile.go [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+		}
+
+		// calculate and update current reconcile status
+		if err := pr.UpdateCurrentReconcileStatusIfNecessary(ctx, p); err != nil {
+			r.Log.Errorf("failed to calculate pipeline status(auto retry), pipelineID: %d, err: %v", p.ID, err)
+			return rutil.ContinueWorkingWithDefaultInterval
+		}
+
+		// check pipeline reconcile done
+		done := !pr.NeedReconcile(ctx, p)
```
Comment (kakj-go) [inline modules/pipeline/providers/reconciler/reconcile.go:87]: The return value is inconsistent with the method name
and then ! Make it difficult to read

File: modules/pipeline/providers/reconciler/taskrun/taskop/queue.go [hunk 9]
```diff
@@ -53,9 +53,9 @@ func (q *queue) Processing() (interface{}, error) {
 		case <-q.Ctx.Done():
 			q.StopQueueLoop = true
 			return
-		case <-q.PExitCh:
-			logrus.Warnf("reconciler: pipeline exit, stop queue, pipelineID: %d, taskID: %d", q.P.ID, q.Task.ID)
-			return
+		//case <-q.PExitCh:
+		//	logrus.Warnf("reconciler: pipeline exit, stop queue, pipelineID: %d, taskID: %d", q.P.ID, q.Task.ID)
+		//	return
```
Comment (Effet) [inline modules/pipeline/providers/reconciler/taskrun/taskop/queue.go]: useless code should be removed

File: modules/pipeline/providers/reconciler/rutil/continue_working.go (hunk 5 above)
Comment (kakj-go) [inline modules/pipeline/providers/reconciler/rutil/continue_working.go:28]: Why use negative numbers ?

File: modules/pipeline/providers/reconciler/util.go [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+			p, exist, err := r.dbClient.GetPipelineWithExistInfo(pipelineID)
+			if !exist {
+				r.Log.Errorf("failed to fetch pipeline detail(no retry), pipelineID: %d, err: %v", pipelineID, dbclient.ErrRecordNotFound)
+				return nil
+			}
+			if err != nil {
+				r.Log.Errorf("failed to fetch pipeline detail(auto retry), pipelineID: %d, err: %v", pipelineID, err)
+				time.Sleep(r.Cfg.RetryInterval)
+				continue
+			}
```
Comment (Effet) [inline modules/pipeline/providers/reconciler/util.go:41]: I think error check should before exist

**Review by Effet — CHANGES_REQUESTED** [review]

File: modules/pipeline/providers/cron/compensator/provider.go [hunk 11]
```diff
@@ -234,15 +278,15 @@ func (p *provider) cronInterruptCompensate(pc spec.PipelineCron) error {
...
+	if err := p.dbClient.UpdatePipelineCron(pc.ID, &pc); err != nil {
 		return errors.Errorf("failed to update pipelineCron for lastCompensateAt field, err: %v", err)
 	}
 
 	return nil
 }
 
-func (p *provider) CronNotExecuteCompensateById(id uint64) error {
-	cron, err := p.client.GetPipelineCron(id)
+func (p *provider) doNotExecuteCompensateByCronID(id uint64) error {
```
[the remark this answers is not in the document]
Comment (sfwn) [reply modules/pipeline/providers/cron/compensator/provider.go:288]: changed to `doNonExecuteCompensateByCronID`

File: modules/pipeline/providers/reconciler/task_reconciler.go (hunk 1 above)
Comment (sfwn) [reply modules/pipeline/providers/reconciler/task_reconciler.go:45]: adjusted to `IdempotentSaveTask`

File: modules/pipeline/providers/reconciler/taskrun/taskrun.go [hunk 12]
```diff
@@ -39,20 +38,19 @@ type TaskRun struct {
...
 	QuitQueueTimeout bool
 	QuitWaitTimeout  bool
 
 	StopQueueLoop bool
 	StopWaitLoop  bool
 
-	PExitCh        <-chan struct{}
-	PExitChCancel  context.CancelFunc
-	PExit          bool
+	//PExitCh        <-chan struct{}
```
[the remark this answers is not in the document]
Comment (sfwn) [reply modules/pipeline/providers/reconciler/taskrun/taskrun.go]: removed

File: modules/pipeline/providers/reconciler/taskrun/taskop/wait.go [hunk 13]
```diff
@@ -71,9 +71,9 @@ func (w *wait) Processing() (interface{}, error) {
 			return executorData.Data, nil
 		case <-w.Ctx.Done():
 			return data, nil
-		case <-w.PExitCh:
-			logrus.Warnf("reconciler: pipeline exit, stop wait, pipelineID: %d, taskID: %d", w.P.ID, w.Task.ID)
-			return data, nil
+		//case <-w.PExitCh:
```
[the remark this answers is not in the document]
Comment (sfwn) [reply modules/pipeline/providers/reconciler/taskrun/taskop/wait.go]: removed

File: modules/pipeline/providers/reconciler/taskrun/taskop/queue.go [hunk 14]
```diff
@@ -53,9 +53,9 @@ func (q *queue) Processing() (interface{}, error) {
 		case <-q.Ctx.Done():
 			q.StopQueueLoop = true
 			return
-		case <-q.PExitCh:
-			logrus.Warnf("reconciler: pipeline exit, stop queue, pipelineID: %d, taskID: %d", q.P.ID, q.Task.ID)
-			return
+		//case <-q.PExitCh:
```
[the remark this answers is not in the document]
Comment (sfwn) [reply modules/pipeline/providers/reconciler/taskrun/taskop/queue.go]: removed

File: modules/pipeline/providers/reconciler/taskpolicy/provider.go [hunk 15]
```diff
@@ -12,61 +12,54 @@
...
+	p.supportedPolicies = map[apistructs.PolicyType]Interface{
+		apistructs.TryLatestSuccessResultPolicyType: tryLastSuccessResult{p: p},
+		apistructs.NewRunPolicyType:                 newRun{p: p},
+		apistructs.TryLatestResultPolicyType:        tryLastResult{},
+	}
+	return nil
 }
 
-func resetTaskByStatusArrayPipeline(statusArray []string, task *spec.PipelineTask, opt PolicyHandlerOptions) (*spec.PipelineTask, error) {
+func (p *provider) resetTaskByStatusArrayPipeline(statusArray []string, task *spec.PipelineTask) error {
```
[the remark this answers is not in the document]
Comment (sfwn) [reply modules/pipeline/providers/reconciler/taskpolicy/provider.go:51]: adjusted to `resetTask(task *spec.PipelineTask, statuses ...apistructs.PipelineStatus)`

File: modules/pipeline/providers/reconciler/task_reconciler.go [hunk 16]
```diff
@@ -0,0 +1,391 @@
...
+	TeardownAfterReconcileDone(ctx context.Context, p *spec.Pipeline, task *spec.PipelineTask)
+	CreateSnippetPipeline(ctx context.Context, p *spec.Pipeline, task *spec.PipelineTask) (snippetPipeline *spec.Pipeline, err error)
+	PrepareBeforeReconcileSnippetPipeline(ctx context.Context, snippetPipeline *spec.Pipeline, snippetTask *spec.PipelineTask) error
+}
+
+type defaultTaskReconciler struct {
+	log    logs.Logger
+	policy taskpolicy.Interface
+
+	pr *defaultPipelineReconciler
```
[the remark this answers is not in the document]
Comment (sfwn) [reply modules/pipeline/providers/reconciler/task_reconciler.go:58]: tr add needed fields.

File: modules/pipeline/providers/reconciler/reconcile.go (hunk 8 above)
Comment (sfwn) [reply modules/pipeline/providers/reconciler/reconcile.go:87]: add interface func `IsReconcileDone`.

File: modules/pipeline/providers/reconciler/rutil/continue_working.go (hunk 5 above)
Comment (sfwn) [reply modules/pipeline/providers/reconciler/rutil/continue_working.go:28]: add comment 'negative numbers are reserved for loop control.'

File: modules/pipeline/providers/reconciler/cache/impl.go (hunk 2 above)
Comment (sfwn) [reply modules/pipeline/providers/reconciler/cache/impl.go:32]: it's a basic key for other typed keys.

File: modules/pipeline/providers/reconciler/pipeline_reconciler.go (hunk 3 above)
Comment (sfwn) [reply modules/pipeline/providers/reconciler/pipeline_reconciler.go:63]: grouped

File: modules/pipeline/providers/reconciler/reconcile.go (hunk 4 above)
Comment (sfwn) [reply modules/pipeline/providers/reconciler/reconcile.go:37]: It makes sense.
defer recover at the beginning of the file.
defer cancel direct after declaration.

File: modules/pipeline/providers/reconciler/rutil/continue_working.go (hunk 5 above)
Comment (sfwn) [reply modules/pipeline/providers/reconciler/rutil/continue_working.go:28]: adjusted to `ContinueWorkingAbort`.

File: modules/pipeline/providers/reconciler/task_reconciler.go (hunk 6 above)
Comment (sfwn) [reply modules/pipeline/providers/reconciler/task_reconciler.go:182]: taskrun pkg will be refactored totally later.

File: modules/pipeline/providers/cron/compensator/provider.go (hunk 7 above)
Comment (sfwn) [reply modules/pipeline/providers/cron/compensator/provider.go:189]: adjusted in another pr.

File: modules/pipeline/providers/reconciler/taskrun/taskop/queue.go (hunk 9 above)
Comment (sfwn) [reply modules/pipeline/providers/reconciler/taskrun/taskop/queue.go]: yes

File: modules/pipeline/providers/reconciler/util.go (hunk 10 above)
Comment (sfwn) [reply modules/pipeline/providers/reconciler/util.go:41]: yes, you are right

**Review by Effet — APPROVED** [review]

**Pull request merged**